### PR TITLE
[Discover-next] Add search bar extensions

### PR DESCRIPTION
**Pull request by joshuali925**
### Description

A search bar extension can display a UI component above the query bar. The component has the ability to read and write discover search bar states to enhance the search experience for users. The configuration is part of Query Enhancements.

```ts
export interface SearchBarExtensionDependencies {
  /**
   * Currently selected index patterns.
   */
  indexPatterns?: IIndexPattern[];
}

export interface SearchBarExtensionConfig {
  /**
   * The id for the search bar extension.
   */
  id: string;
  /**
   * Lower order indicates higher position on UI.
   */
  order: number;
  /**
   * A function that determines if the search bar extension is enabled and should be rendered on UI.
   * @returns whether the extension is enabled.
   */
  isEnabled: () => Promise<boolean>;
  /**
   * A function that returns the mount point for the search bar extension.
   * @param dependencies - The dependencies required for the extension.
   * @returns The mount point for the search bar extension.
   */
  getComponent: (dependencies: SearchBarExtensionDependencies) => React.ReactElement;
}

export interface QueryEnhancement {
  ...
  searchBar?: {
    ...
    extensions?: SearchBarExtensionConfig[];
  };
}
```

### Issues Resolved

Developers can utilize search bar extensions to add additional features to the search bar, such as query assist.

Part of #6820 

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/data/public/ui/search_bar/search_bar.tsx [hunk 1]
```diff
@@ -239,6 +243,15 @@ class SearchBarUI extends Component<SearchBarProps, State> {
     );
   }
 
+  private shouldRenderExtensions() {
```
Comment (kavilla) [inline src/plugins/data/public/ui/search_bar/search_bar.tsx:246]: nice!

File: src/plugins/data/public/ui/search_bar/search_bar.tsx [hunk 2]
```diff
@@ -512,6 +525,20 @@ class SearchBarUI extends Component<SearchBarProps, State> {
           filterBar={filterBar}
           dataTestSubj={this.props.dataTestSubj}
           indicateNoData={this.props.indicateNoData}
+          queryEditorRef={this.queryEditorRef}
+        />
+      );
+    }
+
+    let searchBarExtensions;
+    if (this.shouldRenderExtensions() && this.queryEditorRef.current) {
```
Comment (kavilla) [inline src/plugins/data/public/ui/search_bar/search_bar.tsx:534]: nit: do you think it's worth to shove this into the query editor top row?

File: src/plugins/data/public/ui/search_bar/search_bar.tsx [hunk 3]
```diff
@@ -521,6 +548,7 @@ class SearchBarUI extends Component<SearchBarProps, State> {
     return (
       <div className={className} data-test-subj="globalQueryBar">
         {queryBar}
+        {searchBarExtensions}
```
Comment (kavilla) [inline src/plugins/data/public/ui/search_bar/search_bar.tsx:551]: along with: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6894/files#r1626586752, then we just need to do the work in the query editor

File: src/plugins/data/public/index.ts [hunk 4]
```diff
@@ -527,3 +527,6 @@ export {
   DataSourceGroup,
   DataSourceOption,
 } from './data_sources/datasource_selector';
+
+export { SuggestionsComponent } from './ui';
```
Comment (kavilla) [inline src/plugins/data/public/index.ts:531]: might be related to another PR. but for this could this be just apart of the ui service: https://github.com/opensearch-project/OpenSearch-Dashboards/blob/830664c9298fec49ef73f50596089e7bfaab4fde/src/plugins/data/public/ui/ui_service.ts#L64

if it's not already accessible?

File: src/plugins/data/public/ui/query_editor/query_editor.tsx [hunk 5]
```diff
@@ -521,6 +519,7 @@ export default class QueryEditorUI extends Component<Props, State> {
               </EuiFlexGroup>
             </EuiFlexItem>
             <EuiFlexItem onClick={this.onClickInput} grow={true}>
+              <div ref={this.props.queryEditorRef} />
```
Comment (kavilla) [inline src/plugins/data/public/ui/query_editor/query_editor.tsx:522]: nit: can this be renamed to like queryEditorHeaderRef? to avoid confusion. I'm fine with this name as well since we can think of this and the code editor the same items essentially.

Comment (kavilla) [reply src/plugins/data/public/ui/query_editor/query_editor.tsx:522]: do we want to add a class and styling on this div as well?

two things we should consider adding is a max-height and overflow scroll to restrict how long this could get based what is being portal'd in

**Review by kavilla — APPROVED** [review]

**Comment by kavilla** [comment]
Offline. Will merge as is and then do a fast follow

**Comment by joshuali925** [comment]
thanks, most comments were addressed in https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6895/commits/9c091358e9dc2cada9797ab6298c4fbb95e22f75, any remaining ones i'll open a PR today
